File: haskell/glue/src/Glue/Lib/Bool/If.hs
module Glue.Lib.Bool.If where

import Glue.Eval (Eval, eval, throwError)
import Glue.Eval.Exception (wrongArgumentType)
import Glue.IR (IR (..), isTruthy)

if_ :: IR Eval
if_ = Special ifImpl

ifImpl :: [IR Eval] -> Eval (IR Eval)
ifImpl [cond, thenExpr, elseExpr] = do
    condVal <- eval cond
    eval if isTruthy condVal then thenExpr else elseExpr
ifImpl _ = throwError $ wrongArgumentType ["condition", "then", "else"]

File: dart/glue/lib/src/lib/bool/if.dart
import 'package:glue/src/eval.dart';
import 'package:glue/src/eval/exception.dart';
import 'package:glue/src/ir.dart';

/// If special form
/// Mirrors Haskell Glue.Lib.Bool.If.if_ exactly
final Ir if_ = IrSpecial(ifImpl);

/// If special form implementation
/// Mirrors Haskell Glue.Lib.Bool.If.ifImpl exactly
Eval<Ir> ifImpl(List<Ir> args) {
  return switch (args) {
    [final cond, final thenExpr, final elseExpr] => eval(cond).bind((condVal) {
      return switch (condVal) {
        IrBool(value: false) => eval(elseExpr),
        _ => eval(thenExpr),
      };
    }),
    _ => throwError(wrongArgumentType(['condition', 'then', 'else'])),
  };
}
